design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,789 @@
+- apiGroups:
+  - apiextensions.crossplane.io
+  - pkg.crossplane.io
+  resources: ["*"]
+  verbs: [get, list, watch]
+# Crossplane viewers have read-only access to all of the composite resources and
+# claims it defines.
+- apiGroups: [xr.example.org]
+  resources:
+  - examplecomposites
+  - examplecomposites/status
+  - exampleclaims
+  - exampleclaims/status
+  verbs: [get, list, watch]
+# Crossplane viewers have read-only access to all resources defined by installed
+# providers.
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - exampleproviderconfigs
+  verbs: [get, list, watch]
+```
+
+The RBAC manager creates `edit` and `view` cluster roles for each namespace.
+These cluster roles are 'namespace aligned' - they are intended to grant access
+to a particular namespace via a `RoleBinding` - but not namespace scoped. This
+is because (while these examples are flattened into single roles) they use role
+aggregation, which is not supported by namespaced RBAC roles.
+
+Maintaining an `edit` and a `view` role for each namespace allows the specific
+access that these roles grant to vary from namespace to namespace. A platform
+operator who has been granted the `crossplane-admin` role may influence which
+resource claims are available to a particular namespaces by annotating the
+namespace (see [Composite Resource ClusterRole Mechanics] for details).
+
+> Note that there is no namespace aligned `admin` role. It may be desirable for
+> a subject bound to the `admin` role in a particular namespace to be able to
+> bind other subjects to that role in that namespace, but it is not possible to
+> enforce this. Kubernetes RBAC allows a subject with access to create role
+> bindings to bind any role that grants less or equal access than they have, and
+> thus a subject bound to the `crossplane-ns-example-admin` cluster role in the
+> `example` namespace could bind another subject the `crossplane-ns-other-admin`
+> cluster role in the `example` namespace as long as binding the latter role was
+> not a privilege escalation.
+
+The `crossplane-ns-${namespace}-edit` role grants full access to the resource
+claims that are available to said namespace:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  # This role is intended for use with the 'example' namespace.
+  name: crossplane-ns-example-edit
+rules:
+# Crossplane namespace editors have access to view events.
+- apiGroups: [""]
+  resources: [events]
+  verbs: [get, list, watch]
+# Crossplane namespace editors may need to read or otherwise interact with
+# resource claim connection secrets.
+- apiGroups: [""]
+  resources: [secrets]
+  verbs: ["*"]
+# Crossplane editors have read-only access to composite resource definitions and
+# compositions. This allows them to discover and select an appropriate
+# composition when creating a resource claim.
+- apiGroups: [apiextensions.crossplane.io]
+  resources: ["*"]
+  verbs: [get, list, watch]
```

Comment:
On reflection I'm fairly sure this won't work. XRDs and Compositions are cluster scoped, and I believe rules pertaining to cluster scoped resources are ignored when a `ClusterRole` is bound to a subject in a particular namespace via a `RoleBinding`.

Reply:
@thephred I've introduced a new `crossplane-browse` cluster role to address this. I'm hopeful that this could be integrated into Upbound Cloud either by allowing granting 'browse' (along with admin, edit, or view) at the cluster scope or by automatically granting 'browse' at the cluster scope to anyone granted `edit` at the namespace scope.